rfcs/0032-github-merging-strategy.md

```diff
@@ -0,0 +1,134 @@
+these PR chains correctly.
+
+When multiple PRs modify the same lines of code and one of them is merged into
+`develop`, the other will have 'merge conflicts' and will need manual
+intervention to fix the conflict before it can be merged. We would like to
+minimise this because
+* PRs with conflicts do not merge automatically
+* the PR process is slowed down by waiting for these manual interventions
+* developer time is wasted in resolving these conflicts.
+
+Our current merge strategy is 'squash and merge', which discards the
+individually committed changes within each PR. This hampers `git`'s ability to
+reduce or eliminate merge conflicts by combining the changes from individual
+commits, unnecessarily creating more merge conflicts to be handled manually.
+
+These unnecessary conflicts are particularly prevelant in PR chains. For
+example,
+```
+PR 1
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+
+PR 2 (chained to PR 1)
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Modify line 1 in A.txt]
+-> ...
+```
+when `PR 1` is squashed and merged into develop, `git` is unable to recognise
+that the initial changes made in `PR 2` match with those from `PR 1` and
+reports a merge conflict.
+
+If PRs in chains have review comments that result in changes, these will
+usually also result in a merge conflict. For example,
+```
+PR 1
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Modify line 1 in A.txt in response to review comments]
+
+PR 2 (chained to PR 1)
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Add line 1 in C.txt]
+-> ...
+```
+when `PR 1` is squashed and merged, `git` will again be unable to merge `PR 2`
+automatically.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+We can use `git`'s merge commits when merging to avoid the above issues. These
+preserve the individual commit information that `git` needs to automatically
+resolve these conflicts.
+
+This is a simple configuration change in the GitHub settings for the repo.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+#### We lose a linear `git` history (from `git log` and friends)
+
+We could mitigate this by running a branch in parallel with `develop` that
+'squashes' each merge commit after-the-fact with
+```bash
+git co develop-pretty
+git diff develop~ develop | git apply
+git commit -C develop
+```
+
+Locally, developers can also use `git log -m --first-parent` to view only the
+merge commits. `git log -m --first-parent --patch` shows the changes from
+commits beneath the merge commit, as expected.
+
+#### `git blame` identifies the commit within PRs instead of the merge commit
+
+Similar to the above, developers can locally use `git blame -m --first-parent`,
+or blame on the 'prettified' branch.
+
+#### The history will contain bad or useless commit messages
+
+This can be mostly ignored when using the above mitigations. Changing commit
```

Comment:
My initial reaction is to be not okay with this. But given we can improve prettiness via the side branch and log filtering, it actually seems okay...

Reply:
How often do you actually inspect commit messages?

It's worth mentioning that, should we choose to enforce that PRs are *only* merged through this method (and never manually, except by trusted admins), we can enforce a format for the PR merge commits, which would keep it easy to still see history roughly the way we do today with some simple filtering of the git log.

Reply:
> How often do you actually inspect commit messages?

Whenever I'm bisecting / git blaming . Actually happens fairly often! But happy with `git log` on the other branch or using first-parent